hail/src/main/scala/is/hail/lir/SplitMethod.scala

```diff
@@ -0,0 +1,204 @@
+      def splitLargeStatement(c: StmtX): Unit = {
+        def visit(x: ValueX): Int = {
+          // FIXME this doesn't handle many moderate-sized children
+          val size = 1 + x.children.map(visit).sum
+
+          if (size > SplitMethod.TargetMethodSize / 2) {
+            val l = m.newLocal("spill_large_expr", x.ti)
+            x.replace(load(l))
+            c.insertBefore(store(l, x))
+            1
+          } else
+            size
+        }
+
+        c.children.foreach(visit)
+      }
+
+      var x = b.first
+      while (x != null) {
+        splitLargeStatement(x)
+        x = x.next
+      }
+    }
+  }
+
+  def spillLocals(): Unit = {
+    val (locals, localIdx) = m.findAndIndexLocals(blocks)
+
+    val fields = locals.map { l =>
+      if (l.isInstanceOf[Parameter])
+        null
+      else
+        c.newField(genName("f", l.name), l.ti)
+    }
+
+    def localField(l: Local): Field =
+      l match {
+        case p: Parameter =>
+          if (p.i == 0)
+            null
+          else
+            paramFields(p.i - 1)
+        case _ => fields(localIdx(l))
+      }
+
+    def spill(x: X): Unit = {
+      x.children.foreach(spill)
+      x match {
+        case x: LoadX =>
+          val f = localField(x.l)
+          if (f != null)
+            x.replace(getField(f, load(m.getParam(0))))
+        case x: IincX =>
+          val f = localField(x.l)
+          assert(f != null)
+          x.replace(
+            putField(f, load(m.getParam(0)),
+              insn(IADD,
+                getField(f, load(m.getParam(0))),
+                ldcInsn(x.i, IntInfo))))
+        case x: StoreX =>
+          val f = localField(x.l)
+          assert(f != null)
+          val v = x.children(0)
+          v.remove()
+          x.replace(putField(f, load(m.getParam(0)), v))
+        case _ =>
+      }
+    }
+
+    for (b <- blocks) {
+      var x = b.first
+      while (x != null) {
+        val n = x.next
+        spill(x)
+        x = n
+      }
+    }
+  }
+
+  def splitBlock(b: Block): Unit = {
+    val last = b.last
+
+    val returnTI = last match {
+      case _: GotoX => UnitInfo
+      case _: IfX => BooleanInfo
+      case _: SwitchX => IntInfo
+      case _: ReturnX => m.returnTypeInfo
+      case _: ThrowX => UnitInfo
+    }
+
+    var L = new Block()
+    var x = b.first
+    var size = 0
+
+    while (x != last) {
+      if (size > SplitMethod.TargetMethodSize) {
+        val newM = c.newMethod(genName("m", "wrapped"), FastIndexedSeq.empty[TypeInfo[_]], UnitInfo)
+        L.method = newM
+        newM.setEntry(L)
+        L.append(returnx())
+
+        x.insertBefore(methodStmt(INVOKEVIRTUAL, newM, Array(load(m.getParam(0)))))
+
+        L = new Block()
+        size = 0
+      }
+
+      size += x.approxByteCodeSize()
+      val n = x.next
+      x.remove()
+      L.append(x)
+      x = n
+    }
+
+    val newM = c.newMethod(genName("m", "wrapped"), FastIndexedSeq.empty[TypeInfo[_]], returnTI)
+    L.method = newM
```

Comment:
This (I think) ends up wrapping every block, no matter how small. It looks like we could also choose to move `L` back into `b` if `size` is below some threshold. Do you think it matters?

Reply:
I just wrote the simplest thing to get it to work.  I have some more complicated plans that I think will subsume this.  Basically, I want to peel off structured chunks of code below the target size, update our estimate of the current function, and stop as soon as the current function is below the size.  But we should add this too.  I'll do it in the sequel.

I have splitting working in the stacked PR on this branch.  I ran the performance tests, and it definitely got much worse for large functions/deep loops:

```
                                         sample_qc    2426.5%     29.964    727.083
                             variant_and_sample_qc    1881.2%     39.905    750.713
                                 genetics_pipeline    1425.9%    108.001   1539.998
                         linear_regression_rows_nd     515.4%    212.164   1093.535
...
Geometric mean: 117.9%
Median:  102.8%
```

So we're going to need a slightly better split strategy before that can go in.  (Also, one test that creates 1,200 SeqOps is also failing with method too large.  The skeleton of the control flow is still too large.)